rigging/connection.py

```diff
@@ -1,79 +1,136 @@
-        """
-        return {
-            'command': instruction,
-            'rig_name': self.name,
-        }
+        try:
+            ret = method(dbus_interface="com.redhat.RigInterface")
+            return RigDBusMessage(**ret)
+
+        except dbus.exceptions.DBusException as exc:
+            if exc.get_dbus_name() == "org.freedesktop.DBus.Error.UnknownMethod":
+                raise DBusMethodDoesntExistError(f"{command.name}()")
+            raise exc
+
+        except Exception as exc:
+            print(f"Exception caught while calling {command} on {self.name}:{exc}")
+
+        return None
 
     def destroy_rig(self):
         """
         Instruct the rig to self-terminate without triggering any configured
         actions or generating an archive.
         """
-        return self._communicate(self._create_instruction('destroy'))
+
+        return self._communicate(RigDBusCommand("destroy"))
+
+
+class RigDBusListener(dbus.service.Object):
+
+    _command_map = None
+
+    def __init__(self, rig_name, logger):
+        self._command_map = {}
+        self.logger = logger
+
+        dbus.mainloop.glib.DBusGMainLoop(set_as_default=True)
+        self._bus = dbus.SessionBus()
+        bus_path = f"com.redhat.Rig.{rig_name}"
+        if bus_path in self._bus.list_names():
+            raise DBusServiceExistsError(bus_path)
+        self._bus_name = dbus.service.BusName(
+                            f"com.redhat.Rig.{rig_name}", self._bus,
+                            allow_replacement=False, replace_existing=False)
+        self._loop = GLib.MainLoop()
+        super().__init__(self._bus, f"/RigControl")
+
+    def map_rig_command(self, command_name, callback):
+        self._command_map[command_name] = callback
+
+    def run_listener(self):
+        self._loop.run()
+
+    @dbus.service.method("com.redhat.RigInterface",
+                        in_signature='', out_signature='a{ss}',
+                        async_callbacks=('ok', 'err'))
+    def destroy(self, ok, err):
+        try:
+            _func = self._command_map["destroy"]
+            if not _func:
+                err(RigDBusMessage("Command `destroy` not implemented.",
+                    False).serialize())
+
+            self.logger.info("Destroying rig")
+
+            # this callback is expected to terminate the process, so we need
+            # to send feedback to the client before the function is called.
+            # Using async_callbacks for this, as otherwise the result would
+            # need to be returned from this method, and that will never
+            # happen if _func() performs thread cleanup and finishes the
+            # process.
+            ok(RigDBusMessage("destroyed.", True).serialize())
+            _func()
+
+        except KeyError:
+            self.logger.error(f"Command `destroy` is not defined.")
+        except Exception as exc:
+            self.logger.error(f"Error when trying to destroy rig: {exc}")
```

Comment:
If I'm understanding this correctly, shouldn't we also be leveraging the `err` callback here?

Reply:
Fixed in 3214d50caa55d84cfd968130b7fa4540035e5b3f